pkg/clustermesh/endpointslicesync/endpointslicesync.go

```diff
@@ -0,0 +1,88 @@
+	"github.com/cilium/hive/cell"
+	"k8s.io/client-go/informers"
+	cache "k8s.io/client-go/tools/cache"
+
+	"github.com/cilium/cilium/pkg/clustermesh/operator"
+)
+
+type endpointSliceSync struct {
+	context       context.Context
+	contextCancel context.CancelFunc
+	Metrics       Metrics
+	cm            operator.ClusterMesh
+
+	concurrentClusterMeshEndpointSync int
+
+	meshPodInformer     *meshPodInformer
+	meshNodeInformer    *meshNodeInformer
+	meshServiceInformer *meshServiceInformer
+
+	endpointSliceMeshController  *endpointslice.Controller
+	endpointSliceInformerFactory informers.SharedInformerFactory
+}
+
+func newEndpointSliceSync(lc cell.Lifecycle, params endpointSliceSyncParams) *endpointSliceSync {
+	if !params.Clientset.IsEnabled() || params.ClusterMeshConfig == "" || !params.CfgClusterMesh.ClusterMeshEnableEndpointSync {
+		return nil
+	}
+
+	log.Info("Endpoint Slice Cluster Mesh synchronization enabled")
+
+	epSync := endpointSliceSync{
+		Metrics:                           params.Metrics,
+		cm:                                params.ClusterMesh,
+		concurrentClusterMeshEndpointSync: params.Cfg.ClusterMeshMaxEndpointsPerSlice,
+	}
+	epSync.context, epSync.contextCancel = context.WithCancel(context.Background())
+	epSync.meshPodInformer = newMeshPodInformer(epSync.cm.GlobalServices())
+	epSync.cm.RegisterClusterServiceUpdateHook(epSync.meshPodInformer.onClusterServiceUpdate)
+	epSync.cm.RegisterClusterServiceDeleteHook(epSync.meshPodInformer.onClusterServiceDelete)
+	epSync.meshNodeInformer = newMeshNodeInformer()
+	epSync.cm.RegisterClusterAddHook(epSync.meshNodeInformer.onClusterAdd)
+	epSync.cm.RegisterClusterDeleteHook(epSync.meshNodeInformer.onClusterDelete)
+	epSync.endpointSliceMeshController, epSync.meshServiceInformer, epSync.endpointSliceInformerFactory = newEndpointSliceMeshController(
+		epSync.context, params.Cfg, epSync.meshPodInformer,
+		epSync.meshNodeInformer, params.Clientset,
+		params.Services, epSync.cm.GlobalServices(),
+	)
+	lc.Append(&epSync)
+	return &epSync
+}
+
+func (epSync *endpointSliceSync) Start(startCtx cell.HookContext) error {
+	log.Info("Bootstrap clustermesh EndpointSlice controller")
+
+	epSync.endpointSliceInformerFactory.Start(epSync.context.Done())
+	if err := epSync.meshServiceInformer.Start(epSync.context); err != nil {
+		return err
+	}
+	epSync.endpointSliceInformerFactory.WaitForCacheSync(startCtx.Done())
+
+	if !cache.WaitForCacheSync(startCtx.Done(), epSync.meshServiceInformer.HasSynced) {
+		return fmt.Errorf("waitForCacheSync on service informer not successful")
+	}
+
+	go func() {
+		if err := epSync.cm.ServicesSynced(epSync.context); err != nil {
+			return // The parent context expired, and we are already terminating
+		}
+		epSync.endpointSliceMeshController.Run(epSync.context, epSync.concurrentClusterMeshEndpointSync)
+	}()
+
+	return nil
+}
+
+func (epSync *endpointSliceSync) Stop(cell.HookContext) error {
+	epSync.contextCancel()
+	return nil
+}
```

Comment:
I'd recommend to use `job.Group` instead of raw Start/Stop hook and adhoc goroutine.

Example:
https://github.com/cilium/cilium/blob/2bf8ae746efe113d36811ae36c14e96e527592cb/pkg/ciliumenvoyconfig/cell.go

Reply:
TIL that this exists, thanks! Something like my last push would work then?

Reply:
Checked and look like this run, marking this as resolve but don't hesitate if you see something else!